mumd/src/network/tcp.rs

```diff
@@ -103,7 +103,7 @@ pub async fn handle(
         .await?;
 
         // Handshake (omitting `Version` message for brevity)
-        let state_lock = state.lock().await;
+        let state_lock = state.read().unwrap();
         let username = state_lock.username().unwrap().to_string();
         let password = state_lock.password().map(|x| x.to_string());
         authenticate(&mut sink, username, password).await?;
```

Comment:
There is a potential risk of a deadlock here. If the lock is acquired and the call to authenticate makes the runtime switch context to another task that also wants to acquire the lock, then the process could be blocked and not be able to resolve. To fix this, you simply need to make sure to drop the lock before awaiting it, and reacquiring it after awaiting.

Reply:
Do you think is better to have a drop call?
```
let state_lock = state.read().unwrap();
let username = state_lock.username().unwrap().to_string();
let password = state_lock.password().map(|x| x.to_string());
drop(state_lock);
```
Or a change context?
```
let (username, password) = {
    let state_lock = state.read().unwrap();
    (state_lock.username().unwrap().to_string(),
        state_lock.password().map(|x| x.to_string()))
};
```
First one seems more likely to induce a mistake that cause a deadlock.

Reply:
I think I prefer the second alternative, but both seem fine to me.